### PR TITLE
Fixes Speech Being Audible in Space

### LINKED ISSUE #6124: speech can still be heard in space
Before saycode changes, all dialogue in atmos less areas had to be done through comms. Now atmos has no effect on speech. Don't know if it's a feature or bug.

### LINKED ISSUE #14385: TALKING IN SPACE
#### Description of issue

YOU CAN SPEAK IN SPACE. YOU CAN HEAR IN SPACE.
ANYONE IN HEARING (READ: SIGHT) RANGE OF YOU CAN HEAR YOU WHEN YOU SPEAK.
WE'RE PROBABLY THE ONLY SERVER THAT 'FEATURES' THIS NONSENSE.

#### Difference between expected and actual behavior

SPEAK IN SPACE -> NOBODY CAN HEAR YOU BECAUSE THERE'S NO WAY FOR THE SOUND TO TRAVEL TO ANOTHER PERSON EXCEPT FOR MAYBE THE SPACEMAN NEXT TO YOU IF YOU HOLD HIS HAND OR BUMP HELMETS OR SOMETHING
VS
SPEAK IN SPACE -> THE SPACEMAN NEXT TO YOU, CLOWN FLYING AWAY 7 TILES AWAY FROM YOU, AND PERSON INSIDE THE STATION OPPOSITE OF YOU CAN ALL HEAR YOU

#### Steps to reproduce 

SPEAK
IN SPACE

#### Length of time in which bug has been known to occur

I actually remember this not being the case at some point in the last two years.
Since goonchat, maybe?

#### Server revision

Game: a357ed3737a63af94f833447b62ef44266882e34 
Patches: c59333d8d4afdada839834747f67f85c9c964805

### PR DESCRIPTION
**Pull request by Shadowmech88**
Fixes #6124.
Fixes #14385.
Resurrects #19590.

The PR remains mostly unchanged from the last one. The difference is that with this PR, crewmembers are audible in any conditions while standing next to them. Lip reading or something, I don't know.
If someone speaks within your view but you are too far away to hear them at that pressure, you will see a message like this:
image
Additionally, you will still see the speech bubble as well.

Also, there was a concern raised in the last PR over locations with their own atmosphere such as inflatable shelters. However, I don't see why that would be an issue, since even if you're inside an atmosphere, the shelter being on a turf with a vacuum would still logically prevent the sound from leaving the shelter, so a turf check should be sufficient.

:cl:
 * rscadd: In space, no one can hear you talk. Unless you use the radio or stand right next to them, that is. The distance which speech travels now scales with atmospheric pressure.

### CODE REVIEW

**Comment by RealestEstate** [comment]
suggestion for a message for the user when seeing someone speak in a mute environment:
_person_ speaks but you cannot hear them!

**Comment by DrSnips** [comment]
Not sure how this a bug fix the feature didn't exist before.
I also don't see how this improves the game at all. I don't really care about the immersions if it hinders the ability to interact in the multiplayer focused game.

**Comment by Zth--** [comment]
As I comment in #6124 this feature existed but it got removed or bugged or idk

**Comment by ShiftyRail** [comment]
To quote a comment in the previous PR:

>i thought we got over asinine features for the sake of immulsion with stomach code/masks blocking eating

**Comment by Zth--** [comment]
>hinders the ability to interact in a multiplayer focused game

You'll be forced to speak through the radio, making your conversation public for someone to hear. Your conversations in space are no longer private.

**Comment by DrSnips** [comment]
Oh yes this is exactly what I want to my games. Especially on other Z levels without tcomms where this is most likely to come up.

**Comment by Zth--** [comment]
/me points to the south

You still have ways of communicating

**Comment by DrSnips** [comment]
/me mimes that he wants insulated gloves

**Comment by ShiftyRail** [comment]
/me indicates that his immulsions are shattered

**Comment by Probe1** [comment]
Add some immersive user feedback about space being a poor medium to hear in and we're good.

**Comment by brndd** [comment]
>Oh yes this is exactly what I want to my games. Especially on other Z levels without tcomms where this is most likely to come up.

If only there was some sort of a handheld radio subtype that worked even without telecomms.

**Comment by Kammerjunk** [comment]
outside the scope of this PR: short-range comms built into hardsuits

**Comment by DrSnips** [comment]
Sorry but we have to be realistic about it remember? Green radios are outside the helmet so they don't work too :^)
I don't see reason to add tedium where communication matters most. Low pressure to none areas are likely deep space and a breached station. We shouldn't be hindering players in these situations.

**Comment by Kammerjunk** [comment]
reason: it's cool

**Comment by Durgen** [comment]
Will someone in space be able to hear another person on the other side of a window in the station?

**Comment by Dacendeth** [comment]
You can always make custom signal SBR for privacy.

**Comment by Shadowmech88** [comment]
It feels like some of the commenters here may have not read the whole PR body, so I'd like to reiterate:
### You can still hear people in space or low-atmosphere conditions if you are standing next to them, no radio necessary.

**Comment by Shadowmech88** [comment]
Additionally, @DrSnips it is a bug because the feature existed prior to saycode changes that happened in 2015.

**Comment by Shadowmech88** [comment]
Feedback message added. Also, speech bubbles are now visible even if you can't hear what they said due to pressure.
image

One thing to note is that such feedback messages only appear for when living mobs speak, not when atoms like speakers speak. There isn't really an immersive way to notify you that an object has said something that you can't hear, but I don't think it's a big issue anyway, since that object is also unlikely to be frustrated that you can't hear it.

File: code/modules/mob/living/carbon/human/life.dm [hunk 1]
```diff
@@ -104,7 +104,8 @@ var/global/list/organ_damage_overlays = list(
 		update_mutations()
 
 /mob/living/carbon/human/Life()
-
+	if(!key)
+		say("FUCK")
```
Comment (Zth--) [inline code/modules/mob/living/carbon/human/life.dm]: :joy:

File: code/modules/mob/living/say.dm [hunk 2]
```diff
@@ -259,20 +259,29 @@ var/list/department_radio_keys = list(
...
 /mob/living/send_speech(var/datum/speech/speech, var/message_range=7, var/bubble_type) // what is bubble type?
+	var/visual_range = message_range //the range of hearers who can see that something was said, but not hear the message
 	say_testing(src, "/mob/living/send_speech() start, msg = [speech.message]; message_range = [message_range]; language = [speech.language ? speech.language.name : "None"]; speaker = [speech.speaker];")
 	if(isnull(message_range))
 		message_range = 7
 
-	var/list/listeners = get_hearers_in_view(message_range, speech.speaker) | observers
+	message_range = atmospheric_speech(speech,message_range)
+
+	var/list/total_listeners = get_hearers_in_view(visual_range, speech.speaker)
```
Comment (Probe1) [inline code/modules/mob/living/say.dm:269]: side note when are we going to add astar but for voices so you can hear people off screen.

Comment (Shadowmech88) [reply code/modules/mob/living/say.dm:269]: N3X15 is working on some fancy positional audio stuff for SS13, which I imagine might include things like that.

Comment (Probe1) [reply code/modules/mob/living/say.dm:269]: I was kidding around.  That'd be sick.

**Pull request merged**